### PR TITLE
[SPARK-7874][MESOS] Don’t allocate more than spark.cores.max in fine-grained mode.

### PR DESCRIPTION
**Pull request by dragos**
Currently only coarse-grained mode observes `spark.cores.max`. The fine-grained mode should follow the same limit, and not go above the defined maximum cores.

There's some duplication of logic between the coarse-grained and fine-grained mode. I created [SPARK-10444](https://issues.apache.org/jira/browse/SPARK-10444) to fix that.

Rework of #8671

### CODE REVIEW

**Comment by dragos** [comment]
@tnachen @andrewor14 please have a look

**Comment by nraychaudhuri** [comment]
LGTM

**Comment by dragos** [comment]
Up to date with master, which has a fix in #9282 that was blocking people from testing this one.

**Comment by dragos** [comment]
@tnachen can you please have a look? I addressed all your points from #8671

**Comment by andrewor14** [comment]
@dragos can you fix the style here?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 1]
```diff
@@ -210,12 +216,18 @@ private[spark] class MesosSchedulerBackend(
...
+   * requirements.
+   *
+   * @param d Mesos SchedulerDriver to decline offers
+   * @param offers Mesos offers to be considered
+   * @return a pair of Mesos offers and corresponding WorkerOffer that can be used by the
+   *         fine-grained scheduler.
    */
-  override def resourceOffers(d: SchedulerDriver, offers: JList[Offer]) {
-    inClassLoader() {
+  private[spark] def usableWorkerOffers(d: SchedulerDriver,
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: I think we've consistently put the function parameters that are longer than 100 in newlines so far, can you move `d: SchedulerDriver` to the next line?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 2]
```diff
@@ -239,44 +250,69 @@ private[spark] class MesosSchedulerBackend(
...
+      val mem = getResource(o.getResourcesList, "mem")
+      val cpus = getResource(o.getResourcesList, "cpus")
+      val slaveId = o.getSlaveId.getValue
+      val offerAttributes = toAttributeMap(o.getAttributesList)
+
+      // check offers for
+      //  1. Memory requirements
+      //  2. CPU requirements - need at least 1 for executor, 1 for task
+      val meetsMemoryRequirements = mem >= calculateTotalMemory(sc)
+      val meetsCPURequirements = cpus >= (mesosExecutorCores + scheduler.CPUS_PER_TASK)
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: I think technically this is not correct (which exists in code before, not your change). We can still launch new tasks if there is already an executor running without mesosExecutoreCores. Probably fix this in the future.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 3]
```diff
@@ -239,44 +250,69 @@ private[spark] class MesosSchedulerBackend(
...
           o.getSlaveId.getValue,
           o.getHostname,
-          cpus)
+          taskCores.toInt))
+      } else {
+        None
       }
+    }).flatten
+
+    (usableOffers, workerOffers)
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: What happens if we hit the limit and don't run any more tasks on the remaining usableOffers? We need to decline those too right?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 4]
```diff
@@ -304,20 +340,25 @@ private[spark] class MesosSchedulerBackend(
...
-          listenerBus.post(SparkListenerExecutorAdded(System.currentTimeMillis(), slaveId,
-            // TODO: Add support for log urls for Mesos
-            new ExecutorInfo(o.host, o.cores, Map.empty)))
-        )
-        logTrace(s"Launching Mesos tasks on slave '$slaveId', tasks:\n${getTasksSummary(tasks)}")
-        d.launchTasks(Collections.singleton(slaveIdToOffer(slaveId).getId), tasks, filters)
+      mesosTasks.foreach {
+        case (slaveId, tasks) =>
+          // add the cores reserved for each Mesos executor (one per slave)
+          totalCoresAcquired += getResource(tasks.get(0).getExecutor.getResourcesList, "cpus")
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: We should only account for executor resources once per slave, not for every task

**Comment by tnachen** [comment]
@dragos just left some comments on this PR, sorry I wasn't able to take a deep look earlier. I think there are some open issues that needs to be addressed first. Take a look and let me know what you think.

**Comment by dragos** [comment]
Thanks @tnachen, but since this didn't make it in 1.6, and fine-grained mode is slated for removal, I don't think it makes sense to work on it anymore.

**Comment by blbradley** [comment]
@dragos Where can you see that fine-grained mode is slated for removal? All I see is #9795.

**Comment by dragos** [comment]
Discussions on the dev mailing list.

**Comment by blbradley** [comment]
@dragos Link?

**Comment by dragos** [comment]
The discussions started on Nov. 19 2015, "Removing the Mesos fine-grained mode". but I only have it in my inbox. I'm sure there's some archive somewhere, but I don't have a link at hand.

**Pull request closed without merging**